services/vision/mlvision/ml_model.go

```diff
@@ -0,0 +1,210 @@
+	"go.opencensus.io/trace"
+
+	"go.viam.com/rdk/resource"
+	"go.viam.com/rdk/robot"
+	"go.viam.com/rdk/services/mlmodel"
+	"go.viam.com/rdk/services/vision"
+	"go.viam.com/rdk/utils"
+)
+
+var model = resource.DefaultModelFamily.WithModel("mlmodel")
+
+const (
+	// UInt8 is one of the possible input/output types for tensors.
+	UInt8 = "uint8"
+	// Float32 is one of the possible input/output types for tensors.
+	Float32 = "float32"
+	// DefaultOutTensorName is the prefix key given to output tensors in the map
+	// if there is no metadata. (output0, output1, etc.)
+	DefaultOutTensorName = "output"
+)
+
+func init() {
+	resource.RegisterService(vision.API, model, resource.Registration[vision.Service, *MLModelConfig]{
+		DeprecatedRobotConstructor: func(ctx context.Context, r any, c resource.Config, logger golog.Logger) (vision.Service, error) {
+			attrs, err := resource.NativeConfig[*MLModelConfig](c)
+			if err != nil {
+				return nil, err
+			}
+			actualR, err := utils.AssertType[robot.Robot](r)
+			if err != nil {
+				return nil, err
+			}
+			return registerMLModelVisionService(ctx, c.ResourceName().Name, attrs, actualR, logger)
+		},
+	})
+}
+
+// MLModelConfig specifies the parameters needed to turn an ML model into a vision Model.
+type MLModelConfig struct {
+	resource.TriviallyValidateConfig
+	ModelName string `json:"mlmodel_name"`
+}
+
+func registerMLModelVisionService(
+	ctx context.Context,
+	name string,
+	params *MLModelConfig,
+	r robot.Robot,
+	logger golog.Logger,
+) (vision.Service, error) {
+	_, span := trace.StartSpan(ctx, "service::vision::registerMLModelVisionService")
+	defer span.End()
+
+	mlm, err := mlmodel.FromRobot(r, params.ModelName)
+	if err != nil {
+		return nil, err
+	}
+	classifierFunc, err := attemptToBuildClassifier(mlm)
+	if err != nil {
+		logger.Infow("error turning turn ml model into a classifier:", "model", params.ModelName, "error", err)
+	} else {
+		logger.Infof("model %q fulfills a vision service classifier", params.ModelName)
+	}
+	detectorFunc, err := attemptToBuildDetector(mlm)
+	if err != nil {
+		logger.Infow("error turning turn ml model into a detector:", "model", params.ModelName, "error", err)
+	} else {
+		logger.Infof("model %q fulfills a vision service detector", params.ModelName)
+	}
+	segmenter3DFunc, err := attemptToBuild3DSegmenter(mlm)
+	if err != nil {
+		logger.Infow("error turning turn ml model into a 3D segmenter:", "model", params.ModelName, "error", err)
+	} else {
+		logger.Infof("model %q fulfills a vision service 3D segmenter", params.ModelName)
+	}
+	// Don't return a close function, because you don't want to close the underlying ML service
+	return vision.NewService(name, r, nil, classifierFunc, detectorFunc, segmenter3DFunc)
+}
+
+// Unpack output based on expected type and force it into a []float64.
+func unpack(inMap map[string]interface{}, name string) []float64 {
+	var out []float64
+	me := inMap[name]
+	switch v := me.(type) {
+	case []uint8:
+		for _, t := range v {
```

Comment:
you can do `out = make([]float64, 0, len(v))` in these so you can optimize allocs

Reply:
Ok done.